RFC-0009-open-telemetry.md

```diff
@@ -0,0 +1,100 @@
+- After instrumentation Presto starts the span and register with OpenTelemetry SDK.
+- SDK creates context which is the actual association to the flow and attach to the current span(parent).
+- While performing any operations, Presto adds the required attributes and events to the respective span.
+- In case of sub operations (child span), Presto creates child span, extract the parent context and attach to the child span as parent context so that all parent and child spans get connected.
+- After the operation spans will get ended in the order of creation and update the span state.
+- SDK keeps on checking the flush trigger and if it reaches the batch, all those spans got batched and send to backend.
+- Backend is a system to store, analyse and visualize this telemetry data. Common backends include systems like Jaeger, Instana, Grafana stack, etc.
+
+![Context propagation](/RFC-0009-open-telemetry/context-propagation-coordinator-to-worker.png)
+
+Using context propagation, Signals can be correlated with each other, regardless of where they are generated.
+
+Context contains the information for the sending and receiving service, or execution unit, to correlate one signal with another. For example, if service A calls service B, then a span from service A whose ID is in context can be used as the parent span for the next span created in service B.
+
+Propagation is the mechanism that moves context between services and processes. It serializes or deserializes the context object and provides the relevant information to be propagated from one service to another.
+
+Propagation is usually handled by instrumentation libraries and is transparent to the user. In the event that you need to manually propagate context, you can use the Propagators API.
+
+OpenTelemetry maintains several official propagators. The default propagator is using the headers specified by the W3C TraceContext specification.
+- In Presto in areas where REST calls involved, we use the header for context propagation as per the above image.
+
+- Presto Coordinator fetch the current span context and inject as the traceparent http header. Which is then extracted from the Worker side and use to create the child spans with the parent context.
+
+- In some other areas parent context is available in child context and we directly set the parent context in child spans.
+
+
+## [Optional] Other Approaches Considered
+
+Based on the discussion, this may need to be updated with feedback from reviewers.
+
+## Adoption Plan
+
+Presto Open Telemetry can be configured by modifying the values in presto-main/etc/telemetry.properties
+
+```properties
+otel-factory.name=otel
+tracing-enabled=false
+tracing-backend-url=<backend endpoint>
+max-exporter-batch-size=256
+max-queue-size=1024
+schedule-delay=1000
+exporter-timeout=1024
+span-sampling=true
+```
+
+***otel-factory.name***: unique identifier for OpenTelemetry factory implementation to be registered
```

Comment:
This seems to imply that the only tracing providers supported must implement the Open Telemetry spec. The runtime should have no knowledge of any aspect of the Open Telemetry spec.  The Open Telemetry spec should be implemented as a plugin and the SPI integration with the runtime should have a generic concept of tracing without any knowledge of any particular spec.

As an example, the Presto runtime and execution engines do not have any knowledge of the Iceberg spec either.  They communicate through the connector SPI.

Please update your proposal with specifics on the abstractions that will be used within the SPI to accomplish this richer tracing functionality.  Thank you.

Reply:
Hi @tdcmeehan The abstraction in SPI is generic and can be implemented by any tracing providers. If the key with "otel" and "OpenTelemetry factory" in the description are confusing here we could modify to some generic name like "tracing-factory-name" and "factory".

Reply:
Please add details on the changes being made to the SPI in this document.

Reply:
Hi @tdcmeehan Added the SPI details. Please review.

Reply:
I don't see it.  Can you add actual interfaces and update the flow chart to denote runtime/engine -> libray, with the arrow indicating which SPI method will be invoked?

Reply:
Hi @tdcmeehan Added the flow chart. Please check.